pynxtools/dataconverter/readers/xps/vms/vamas.py

```diff
@@ -0,0 +1,1032 @@
+"""
+Class for reading XPS files from raw VMS data.
+"""
+# Copyright The NOMAD Authors.
+#
+# This file is part of NOMAD. See https://nomad-lab.eu for further info.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# pylint: disable=too-many-lines
+
+import re
+from copy import deepcopy
+import datetime
+from abc import ABC, abstractmethod
+from itertools import groupby
+import xarray as xr
+import numpy as np
+
+from pynxtools.dataconverter.readers.xps.vms.vamas_data_model import VamasHeader, Block
+
+from pynxtools.dataconverter.readers.xps.reader_utils import (
+    XPSMapper,
+    construct_entry_name,
+    construct_data_key,
+    construct_detector_data_key,
+)
+
+
+class VamasMapper(XPSMapper):
+    """
+    Class for restructuring .txt data file from
+    Vamas format into python dictionary.
+    """
+
+    config_file = "config_vms.json"
+
+    def __init__(self):
+        self.file = None
+        self.parsers = [
+            VamasParserRegular,
+            VamasParserIrregular,
+        ]
+        self.parser_map = {
+            "regular": VamasParserRegular,
+            "irregular": VamasParserIrregular,
+        }
+
+        super().__init__()
+
+    def _select_parser(self):
+        """
+        Select parser based on the structure of the Vamas file,
+        i.e., whether it is regular or irregular.
+
+        Returns
+        -------
+        VamasParserVMS
+            Vamas parser for reading this file structure.
+
+        """
+        vms_type = self._get_vms_type()
+        return self.parser_map[vms_type]()
+
+    def _get_vms_type(self):
+        """Check if the vamas file is regular or irregular"""
+        contents = []
+        with open(self.file, "rb") as vms_file:
+            for line in vms_file:
+                if line.endswith(b"\r\n"):
+                    contents += [line.decode("utf-8", errors="ignore").strip()]
+
+        for vms_type in self.parser_map:
+            if vms_type.upper() in contents:
+                return vms_type
+        return None
```

Comment:
Try to use the same type of return, it is useful when you varify the output from a file. E.g. if vms type is a string then you can return ''.